sql/src/main/java/io/crate/metadata/Signature.java

```diff
@@ -22,137 +22,363 @@
+         * occurs.
+         *
+         * @param secondary the operator which will be evaluated with this operator's success result
+         * @return a new operator with the secondary appended
+         */
+        default SignatureOperator and(SignatureOperator secondary) {
+            return dataTypes -> {
+                List<DataType> n = this.apply(dataTypes);
+                if (n != null) {
+                    return secondary.apply(n);
+                }
+                return null;
+            };
+        }
+
     }
 
-    public Signature(int varArgsStartingPosition, DataType... elements) {
-        assert varArgsStartingPosition < elements.length : "varArgs starting position exceeds args definition";
-        this.varArgsStartingPosition = varArgsStartingPosition;
-        list = ImmutableList.copyOf(elements);
+    /**
+     * Interface for validating and normalizing an argument type of a function signature.
+     * <p>
+     * This interface is an unary operator for {@DataType} which returns a normalized type if the given input type
+     * is valid. Implementations can either pass through the input or replace the type if normalization is required.
+     * For example if the input type is null implementations might return a specific type if possible.
+     * <p>
+     * The operator returns null if the input does not match its criteria.
+     */
+    public interface ArgMatcher extends UnaryOperator<DataType> {
+
+        ArgMatcher ANY = dt -> dt;
+        ArgMatcher NUMERIC = of(DataTypes.NUMERIC_PRIMITIVE_TYPES::contains);
+
+        ArgMatcher ANY_ARRAY = of(dt -> dt instanceof ArrayType);
+        ArgMatcher ANY_SET = of(dt -> dt instanceof SetType);
+        ArgMatcher ANY_COLLECTION = of(dt -> dt instanceof CollectionType);
+
+        ArgMatcher STRING = rewriteTo(DataTypes.STRING);
+        ArgMatcher BOOLEAN = rewriteTo(DataTypes.BOOLEAN);
+        ArgMatcher INTEGER = rewriteTo(DataTypes.INTEGER);
+        ArgMatcher OBJECT = rewriteTo(DataTypes.OBJECT);
+
+        /**
+         * Creates a matcher witch matches any of the given types or the undefined type.
+         *
+         * @param allowedTypes the allowed types
+         * @return a new argument matcher
+         */
+        static ArgMatcher of(DataType... allowedTypes) {
+            return dt -> {
+                for (DataType allowedType : allowedTypes) {
+                    if (allowedType.equals(dt)) {
+                        return dt;
+                    }
+                }
+                return IS_UNDEFINED.test(dt) ? dt : null;
+            };
+        }
+
+        /**
+         * Creates a matcher which matches if any of the given input predicates matches or the input is of type undefined.
+         *
+         * @param predicates the predicates to match
+         * @return a new argument matcher
+         */
+        @SafeVarargs
+        static ArgMatcher of(final Predicate<DataType>... predicates) {
+            return dt -> {
```

Comment:
I think things like this could be made more readable by splitting it like this:

```
        @SafeVarargs
        static ArgMatcher of(final Predicate<DataType>... predicates) {
            return dt -> firstMatchOrNull(dt, predicates);
        }

        @SafeVarargs
        static DataType firstMatchOrNull(DataType dt, final Predicate<DataType>... predicates) {
            for (Predicate<DataType> predicate : predicates) {
                if (predicate.test(dt)) {
                    return dt;
                }
            }
            return IS_UNDEFINED.test(dt) ? dt : null;
        }
```

But that might be overkill as it is simple enough.